hpcbench/benchmark/minigemm.py

```diff
@@ -0,0 +1,261 @@
+      A[i] = 1.1*(i+1);
+    }
+    for(int i = 0; i < k*n; ++i) {
+      B[i] = 1.2*(i+2);
+    }
+    for(int i = 0; i < m*n; ++i) {
+      C[i] = 0.0;
+    }
+  }
+
+  #pragma omp parallel for schedule(static, 1) num_threads(nthreads)
+  for (int ithrd = 0; ithrd < nthreads; ithrd++) {
+    cblas_dgemm(CblasRowMajor, CblasNoTrans, CblasNoTrans,
+                m, n, k, alpha, As[ithrd], k, Bs[ithrd], n,
+                beta, Cs[ithrd], n);
+  }
+
+  auto tstart = std::chrono::high_resolution_clock::now();
+  #pragma omp parallel for schedule(static, 1) num_threads(nthreads)
+  for (int ithrd = 0; ithrd < nthreads; ithrd++) {
+    for (int iter = 0; iter < niter; iter++) {
+      cblas_dgemm(CblasRowMajor, CblasNoTrans, CblasNoTrans,
+                  m, n, k, alpha, As[ithrd], k, Bs[ithrd], n,
+                  beta, Cs[ithrd], n);
+    }
+  }
+  auto tend = std::chrono::high_resolution_clock::now();
+
+  std::chrono::duration<double> tdiff = tend - tstart;
+  std::cout << "Time elapsed for " << niter << " iterations: "
+            << tdiff.count() << "s\\n";
+  std::cout << gflop/tdiff.count() << " GFLOP/s\\n";
+
+#ifdef DBG_PRINT
+  for(int i = 0; i < min(m, 5); ++i) {
+    for(int j = 0; j < min(n, 5); ++j) {
+      std::cout << Cs[0][j + i*n] << "  ";
+    }
+    std::cout << std::endl;
+  }
+#endif
+
+  double chk;
+  double sgn = 1.0;
+  for(int j = 0; j < nthreads; j++) {
+    for(int i = 0; i < m*n; ++i) {
+      sgn *= -1.0;
+      chk += sgn*Cs[j][i];
+    }
+  }
+  std::cout << "Check value: " << chk << std::endl;
+
+  for(int j = 0; j < 2; j++) {
+    mkl_free(As[j]);
+    mkl_free(Bs[j]);
+    mkl_free(Cs[j]);
+  }
+  free(As);
+  free(Bs);
+  free(Cs);
+
+  return 0;
+}
+"""
+
+COMPILE_SCRIPT = """#!/bin/bash -l
+make DEFINES="{opts}"
+"""
+
+MAKEFILE = """CXX=icpc
+DEFINES=
+CXXFLAGS=-qopt-streaming-stores always -O3 -mkl -std=c++11 -qopenmp
+LDFLAGS=
+
+CPPFILES=minigemm.cpp
+OBJ=$(CPPFILES:.cpp=.o)
+
+all: minigemm
+
+.cpp.o: $<
+	$(CXX) -c $< $(DEFINES) $(CXXFLAGS)
+
+minigemm: $(OBJ)
+	$(CXX) -o $@ $^ $(CXXFLAGS) $(LDFLAGS)
+
+clean:
+	$(RM) minigemm *.o
+
+.PHONY: clean
+
+"""
+
+
+class MiniGEMM(Benchmark):
+    """A mini benchmark performing MKL based generalized matrix multiplies
+    to determine the node's peak floating point performance.
+    """
+    name = 'minigemm'
+    description = "DGEMM mini benchmark"
+    COMPILE_PARAMS = []
+    DEFAULT_OPENMP_THREADS = 1
+    DEFAULT_NESTED_MKL_THREADS = 4
+
+    def __init__(self):
+        super(MiniGEMM, self).__init__(
+              attributes=dict(compile=self.COMPILE_PARAMS,
+                              omp_threads=self.DEFAULT_OPENMP_THREADS,
+                              mkl_threads=self.DEFAULT_NESTED_MKL_THREADS
+        ))
+
+    @cached_property
+    def compile(self):
+        return self.attributes['compile']
+
+    @cached_property
+    def omp_threads(self):
+        return str(self.attributes['omp_threads'])
+
+    @cached_property
```

Comment:
I agree with you. Here it's a bit special because depending on your node HW configuration you will have to parametrize this benchmark slightly differently. we could probably use `lscpu` here to auto-determine the number of cores and the NUMA domains to get it right, but I'd rather have the user configure this.
I will add the docstrings though before merging - I didn't know this, any good example I can use?

Reply:
You can take a look at other benchmark classes to have examples. For good ones, I am not sure :)